ios/sound.mm

```diff
@@ -24,209 +24,275 @@
-  }
-
-  // if the selected driver was not found, return an error message
-  if ( iDriverIdx == INVALID_INDEX )
-  {
-      return tr ( "The current selected audio device is no longer present in the system." );
-  }
-
-  // check device capabilities if it fulfills our requirements
-  const QString strStat = CheckDeviceCapabilities ( iDriverIdx );
-
-  // check if device is capable and if not the same device is used
-  if ( strStat.isEmpty() && ( strCurDevName.compare ( strDriverNames[iDriverIdx] ) != 0 ) )
-  {
-  }
-  // set the left/right in/output channels
-   
-  return strStat;
-   */
-    return "";
+    return noErr;
 }
 
-QString CSound::CheckDeviceCapabilities ( const int iDriverIdx )
+void CSound::processBufferList ( AudioBufferList* inInputData, CSound* pSound ) // got stereo input data
 {
-  /*// This function checks if our required input/output channel
-  // properties are supported by the selected device. If the return
-  // string is empty, the device can be used, otherwise the error
-  // message is returned.
-
-  // check sample rate of input
-  audioSession.setPreferredSampleRate( SYSTEM_SAMPLE_RATE_HZ );
-  // if false, try to set it
-  if ( audioSession.sampleRate != SYSTEM_SAMPLE_RATE_HZ )
-  {
-    throw CGenErr ( tr ("Could not set sample rate. Please close other applications playing audio.") );
-  }
-
-  // special case with 4 input channels: support adding channels
-    if ( ( lNumInChan == 4 ) && bInputChMixingSupported )
-    {
-        // add four mixed channels (i.e. 4 normal, 4 mixed channels)
-        lNumInChanPlusAddChan = 8;
-
-        for ( int iCh = 0; iCh < lNumInChanPlusAddChan; iCh++ )
-        {
-            int iSelCH, iSelAddCH;
+    QMutexLocker locker ( &pSound->MutexAudioProcessCallback );
+    Float32*     pData = static_cast<Float32*> ( inInputData->mBuffers[0].mData );
 
-            GetSelCHAndAddCH ( iCh, lNumInChan, iSelCH, iSelAddCH );
-
-            if ( iSelAddCH >= 0 )
-            {
-                // for mixed channels, show both audio channel names to be mixed
-                channelInputName[iCh] =
-                    channelInputName[iSelCH] + " + " + channelInputName[iSelAddCH];
-            }
-        }
-    }
-    else
+    // copy input data
+    for ( int i = 0; i < pSound->iCoreAudioBufferSizeMono; i++ )
     {
-        // regular case: no mixing input channels used
-        lNumInChanPlusAddChan = lNumInChan;
+        // copy left and right channels separately
+        pSound->vecsTmpAudioSndCrdStereo[2 * i]     = (short) ( pData[2 * i] * _MAXSHORT );     // left
+        pSound->vecsTmpAudioSndCrdStereo[2 * i + 1] = (short) ( pData[2 * i + 1] * _MAXSHORT ); // right
     }
-*/
-    // everything is ok, return empty string for "no error" case
-    return "";
+    pSound->ProcessCallback ( pSound->vecsTmpAudioSndCrdStereo );
 }
 
-void CSound::SetLeftInputChannel  ( const int iNewChan )
+// TODO - CSound::Init is called multiple times at launch to verify device capabilities.
```

Comment:
This smells like maybe the calling code needs to be reworked?

Reply:
Yes, see #1727